paper/automated.tex

```diff
@@ -1,20 +1,108 @@
+The ATP then try to refute the negation by applying rules of the underlying calculus, until they find a proof of false (a contradiction).
+In this case, the conjecture was (classicaly) true, and the ATP has found a proof by contradiction, often called a ``refutation'' or ``saturation'' proof.
+
+The underlying calculus varies from system to system, but they often have a variant of a resolution clause, a clause of a form:
+\[\infer{C \lor D}{C \lor L \quad D \lor \neg L} \]
+This can also be read as $C \lor L \quad D \lor \neg L$ implies $C \lor D$, where $C, D, L$ are formulas in e.g. first-order logic.
+Superposition calculi have a variant of this rule that deals with equality directly, and thus are more efficient at reasoning about equality.
+
+In this project we used Vampire~\cite{DBLP:conf/cav/KovacsV13}, Duper~\cite{DBLP:conf/itp/CluneQBA24} and Prover9 and Mace4~\cite{prover9-mace4} which are all based on variants of saturation for proving.
+
+\paragraph{Equational Reasoning}
+
+Equational reasoning is a type of reasoning based on equational logic and rewriting with congruence~\cite{term-rewriting}, see Chapter~\ref{todo} for a discussion of its foundations in universal algebra.
+In general, it takes a series of equations and determines wether another equation can be deduced from it.
+A core tool in equational reasoning are e-graphs, a data structure used to represent equivalence classes of terms.
+The procedure of congruence closure that can be used to decide ground equational problems (i.e. problems without variables) and as a semi-decision procedure in general.
+
+SMT solvers like Z3~\cite{DBLP:conf/tacas/MouraB08} use equational reasoning for deciding the theory of equality with uninterpreted functions~\cite{DBLP:series/txtcs/KroeningS16,DBLP:conf/cade/MouraB07}.
+On the other hand, equality saturation~\cite{DBLP:journals/pacmpl/WillseyNWFTP21} uses e-graphs by extending congruence closure to a more controlled search, enabling optimization and conditional rewriting.
+One of the main advantages of using equational reasoning to reason about implications of magma laws is that we get very explicit proofs: a proof that $l \vdash l'$ is given by a sequence of rewrites that starts at the left-hand side of $l'$ and arrives at the right-hand side through applications of $l$.
+
+In this project we used Z3~\cite{DBLP:conf/tacas/MouraB08}, Prover9 and Mace4~\cite{prover9-mace4}, a custom ATP for magmas based on egg~\cite{DBLP:journals/pacmpl/WillseyNWFTP21}, and the Lean egg tactic~\cite{DBLP:journals/pacmpl/KoehlerGBGTS24,rossel2024bridging}, which all work with equational logic. We have also reasoned with manual (custom written) heuristics about simple rewrites.
+
+\subsection{Proof Reconstruction and Integration}
+\label{sec:proof-reconstruction}
+
+While ATPs are very useful to solve questions for this project, they generally don't integrate well with Lean (except for Lean tactics like \texttt{duper}, \texttt{aesop} and \texttt{egg}).
```

Comment:
"While ATPs are very useful to solve questions for this project, they generally don't integrate well with Lean" I would be more specific: "ATPs can resolve many implications or non-implications studied by this project; however, except for Lean tactics like ..., ATP systems do not generate proof certificates that are natively recognized by Lean."